change the `consensus::replicate` method to accept `chunked_vector`

mmaslankaprv opened the pull request:
Previously `consensus::replicate` accepted `model::record_batch_reader`
as an argument. This is a bit of an overuse of reader as every time
batches are replicated the code simply wraps a vector of batches with
in memory record batch reader. This commit simplifies the interface and
changes the interface to accept `chunked_vector<model::record_batch>`

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv commented:
/ci-repeat 1

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--8dadba99-d04b-477f-ba25-91f826b4a0ed-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/59472#0193aa52-2d25-45d6-9c80-a3c6a2e77d0e have failed and will be retried
```
idempotency_tests_rpunit
rm_stm_tests_rpunit
```
<!--/body-->

mmaslankaprv commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59850-->
<!--9ad115f9-f547-4f40-b95c-f613b01d2f9c-->
<!--title-->
## Retry command for [Build#59850](https://buildkite.com/redpanda/redpanda/builds/59850)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":true,"with_iceberg":false,"with_tiered_storage":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9d5c1365-fccc-42a8-ab5f-baf214122b4b-->
<!--TEST RESULTS-->
<!--c40ddb28-4979-4817-9733-abeb38d1b4f6-->
<!--TEST RESULTS-->
<!--98e1ace4-1fd5-46b6-8749-0c6feb930d2e-->
<!--TEST RESULTS-->
<!--a1e27e0b-d4bd-40a3-aaf1-e10f2d3c7bb1-->
<!--TEST RESULTS-->
<!--73eda628-527b-4823-af2d-4b7d4c2d8dd2-->
<!--TEST RESULTS-->
<!--c599974a-469b-496a-9d65-d3a9d345647b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59850">build#59850</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/59850#0193d468-b972-41b1-8d7f-aefbb607c811 | FLAKY | 1/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/59850#0193d468-b972-4df1-8338-79914d6d3e28 | FLAKY | 1/2 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59850#0193d4a9-e158-4c45-af30-9c427dc97f01 | FLAKY | 3/6 |
| rptest.tests.datalake.simple_connect_test.RedpandaConnectIcebergTest.test_translating_avro_serialized_records.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59850#0193d4a9-e157-4e6b-85c0-0beff927e245 | FLAKY | 5/6 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59850#0193d4a9-e156-429e-a7b1-10108a3876ff | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59871">build#59871</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/59871#0193d5c4-e938-47c6-8052-a4db4b3e89ba | FLAKY | 1/2 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59871#0193d61d-670d-4f22-b53d-24d778b038f8 | FLAKY | 4/6 |
| rptest.transactions.stream_verifier_test.StreamVerifierTest.test_simple_produce_consume_txn_with_add_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/59871#0193d61d-670f-4ce8-965d-45c168220e3e | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59911">build#59911</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/59911#0193d9d9-2bf4-43da-a193-babf511ca06b | FLAKY | 1/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/59911#0193d9d9-2bf5-41a1-b11c-e7705baa18cd | FLAKY | 1/2 |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59911#0193da36-11bd-44af-ac70-9a28e5e82ac3 | FAIL | 0/6 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitMirrorMakerTests.test_mirror_maker_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/59911#0193da36-11be-497c-979a-7e9729530e32 | FLAKY | 5/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59911#0193da31-f113-40de-86c7-d93da3beae1e | FLAKY | 2/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59911#0193da36-11be-4d7a-b109-ea13a4b3fc38 | FLAKY | 5/6 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_write.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59911#0193da36-11bd-44af-ac70-9a28e5e82ac3 | FLAKY | 3/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60081">build#60081</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60081#0193f47b-54d1-45ea-98e9-7699d1e89e92 | FLAKY | 1/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60081#0193f47b-54d0-42db-a67d-bf2fa9182f46 | FLAKY | 1/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60081#0193f47b-54d1-45ea-98e9-7699d1e89e92 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60213">build#60213</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60213#01942611-9f42-4c90-b499-eda0be6e79e0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60265">build#60265</a></summary>


| test_id                                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| idempotency_tests_rpunit.idempotency_tests_rpunit                                                                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/60265#01942ca9-bcb7-4106-8cee-a4bd20015efd | FLAKY         | 1/2      |
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/60265#01942ca9-bcb7-4106-8cee-a4bd20015efd | FLAKY         | 1/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                                                                              | unit        | https://buildkite.com/redpanda/redpanda/builds/60265#01942ca9-bcb6-4aa7-9aa2-af0903997685 | FLAKY         | 1/2      |
| rptest.tests.datalake.simple_connect_test.RedpandaConnectIcebergTest.test_translating_avro_serialized_records.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60265#01942cee-a5ec-4e7b-a0b3-466ef57468de | FAIL          | 0/1      |
</details>
<!--/body-->

mmaslankaprv commented:
/dt

bashtanov commented on src/v/container/fragmented_vector.h:
```diff
@@ -480,6 +480,15 @@ class fragmented_vector {
     void erase_to_end(const_iterator begin) { pop_back_n(cend() - begin); }
 
+    template<typename Range>
+    requires std::ranges::range<Range>
```
`std::ranges::sized_range` to avoid repeated scanning for the size? And if we need it for these one day another override should be added, probably without `reserve` call.

bashtanov commented on src/v/kafka/data/replicated_partition.cc:
```diff
@@ -354,8 +360,10 @@ raft::replicate_stages replicated_partition::replicate(
     }
-
-    auto res = _partition->replicate_in_stages(batch_id, std::move(rdr), opts);
+    chunked_vector<model::record_batch> batches;
+    batches.push_back(std::move(batch));
```
FYI only: for these situations `std::ranges::single_view` can work with `chunked_vector::from` if you find it neater.

bashtanov commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -177,13 +165,13 @@ static partition_produce_stages partition_append(
   std::chrono::milliseconds timeout_ms) {
     auto stages = partition->replicate(
-      bid, std::move(reader), acks_to_replicate_options(acks, timeout_ms));
+      bid, {std::move(batch)}, acks_to_replicate_options(acks, timeout_ms));
```
AFAIK in initializer lists `std::move` doesn't eliminate the copy. Initializer list contains values and constructor copies from them.

bashtanov commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -437,7 +426,7 @@ static partition_produce_stages produce_topic_partition(
                           std::move(partition)),
                         bid,
-                        std::move(reader).assume_value(),
+                        std::move(batch).assume_value(),
```
Out of scope, but IIRC `result` can contain neither a value nor an error!

bashtanov commented on src/v/kafka/server/tests/fetch_test.cc:
```diff
@@ -153,6 +153,16 @@ SEASTAR_THREAD_TEST_CASE(partition_iterator) {
+    res.reserve(batches.size());
+    std::move(batches.begin(), batches.end(), std::back_inserter(res));
+    return res;
+}
```
doesn't `fragmented_vector<T>::from` do the same?

bashtanov commented on src/v/container/fragmented_vector.h:
```diff
@@ -480,6 +480,15 @@ class fragmented_vector {
 
+    template<typename Range>
+    requires std::ranges::range<Range>
+    static fragmented_vector from(Range range) {
```
What do you think of using `std::from_range_t` constructor parameter instead of a static method? This would keep the API more compatible with `std::vector`.

mmaslankaprv replied:
i will change it for constructor, but `std::from_range_t` is available since C++23

bashtanov reviewed (commented):
There is a typo in the commit message: "conesnsus".

Other suggestions:
- separate infrastructure improvements such as `chunked_vector::from` into a separate commit;
- what do you think about adding `static chunked_vector<T>::single(T &&)` or even `chunked_vector<T>::chunked_vector(std::in_place_t, Args&&...args)` to reduce boilerplate when creating a single-batch chunked_vector?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59911-->
<!--513e646c-d541-4c47-8a04-95a557abf4c4-->
<!--title-->
## Retry command for [Build#59911](https://buildkite.com/redpanda/redpanda/builds/59911)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

bashtanov commented:
As per our slack discussion, all my concerns regarding copying were invalid, I have removed the comments.

ztlpn commented on src/v/raft/tests/raft_group_fixture.h:
```diff
@@ -827,12 +818,13 @@ inline ss::future<bool> replicate_random_batches(
-          auto rdr = random_batches_reader(count);
-          return random_batches_reader(count).then(
-            [&leader_node, c_lvl, expected_term](auto rdr) {
+          auto rdr = random_batches(count);
```
not needed?

ztlpn commented on src/v/raft/tests/append_entries_test.cc:
```diff
@@ -825,16 +823,16 @@ FIXTURE_TEST(test_big_batches_replication, raft_test_fixture) {
 }
 struct request_ordering_test_fixture : public raft_test_fixture {
-    model::record_batch_reader make_indexed_batch_reader(int32_t idx) {
+    chunked_vector<model::record_batch> make_indexed_batch_reader(int32_t idx) {
```
nit: indexed_batch_reader -> indexed_batches / indexed_batch

ztlpn commented on src/v/kafka/server/tests/replicated_partition_test.cc:
```diff
@@ -59,12 +59,12 @@ FIXTURE_TEST(test_replicated_partition_end_offset, redpanda_thread_fixture) {
               ->replicate(
-                model::make_memory_record_batch_reader(
-                  {std::move(builder).build()}),
+                std::move(batches),
```
nit: could just use single-batch interface

ztlpn commented on src/v/cluster/distributed_kv_stm.h:
```diff
@@ -400,11 +400,12 @@ class distributed_kv_stm final : public raft::persisted_stm<> {
         auto r = co_await _raft->replicate(
           _insync_term,
-          std::move(reader),
+          std::move(batches),
```
nit: single-batch interface?

ztlpn commented on src/v/cluster/tests/tx_compaction_utils.h:
```diff
@@ -319,11 +319,11 @@ class tx_executor {
             auto result = co_await _ctx._stm->replicate(
               bid,
-              std::move(reader),
+              std::move(batches_v),
```
nit: single-batch interface?

ztlpn commented on src/v/kafka/server/group.cc:
```diff
@@ -2227,12 +2226,11 @@ group::offset_commit_stages group::store_offsets(offset_commit_request&& r) {
     auto replicate_stages = _partition->raft()->replicate_in_stages(
       _term,
-      std::move(reader),
+      std::move(batches),
```
nit: single-batch interface?

mmaslankaprv committed:
container/f_vec: added convenience interface to create vector

Added constructors and methods to create a fragmented vector from range
and from single element.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/consensus: used chunked_vector in `raft::replicate`

When replicated batches in Redpanda we never use the capabilities of
record batch reader. It introduces additional burden as it needs to be
converted to vector of batches for batching and handling. Changed the
interface to accept `chunked_vector<model::record_batch>` or a single
batch instead of using readers.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

dotnwat commented on src/v/container/fragmented_vector.h:
```diff
@@ -119,6 +120,24 @@ class fragmented_vector {
+    /**
+     * Constructs a vector with single element
+     */
+    template<typename... Args>
+    explicit fragmented_vector(std::in_place_t, Args&&... args) {
+        emplace_back(std::forward<Args>(args)...);
+    }
```
in the interest of making fragmented_vector behave like std::vector, you might consider making this private since this constructor signature doesn't exist with std::vector. i don't know if we are trying to do that or not, actually, but looking at the existing constructors it does seem like that may have been an intention?

mmaslankaprv replied:
i removed the additional constructor completely but left the factory method. It is a convenient interface we already have some additional methods in `fragmented_vector`. I have a plan to revisit `cluster` module APIs to see if we need vector based replication or single batch is enough.

mmaslankaprv replied:
I need to refactor the `consensus` API as it already have 6 `replicate` method overrides

dotnwat commented on src/v/cluster/tests/partition_properties_stm_test.cc:
```diff
@@ -116,10 +116,13 @@ struct partition_properties_stm_fixture : raft::raft_fixture {
+                .then([&](chunked_vector<model::record_batch> batches) {
                     return leader_node.raft()
                       ->replicate(
-                        model::make_memory_record_batch_reader(
-                          std::move(batches)),
```
clarification: is there something that prevented `chunked_vector<model::record_batch>(std::move(batches));` from being used here directly?

dotnwat commented on src/v/cluster/tests/tx_compaction_utils.h:
```diff
@@ -319,11 +319,11 @@ class tx_executor {
               bid,
-              std::move(reader),
+              chunked_vector<model::record_batch>::single(
+                std::move(batches[0])),
```
did you add a single batch interface to replicate so you don't need the ::single factory here?

mmaslankaprv replied:
this is a bit different situation as `replicate` comes from `rm_stm`

dotnwat reviewed (commented):
lgtm. seems like if you added a single batch interface to `replicate_in_stages` you could get rid of the ::single factory since it would only be needed in a couple call sites in consensus where it would only save one line of code.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60265-->
<!--8fa18c94-debb-44ca-9431-9c05128b2ce6-->
<!--title-->
## Retry command for [Build#60265](https://buildkite.com/redpanda/redpanda/builds/60265)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/simple_connect_test.py::RedpandaConnectIcebergTest.test_translating_avro_serialized_records@{"cloud_storage_type":1}
```
<!--/body-->

WillemKauf commented:
Were there any functional changes in this PR?

From https://redpandadata.atlassian.net/browse/CORE-8765, `idempotency_tests_rpunit.idempotency_tests_rpunit` has seen a large uptick in failures in the past week.
